### PR TITLE
feat: csv export for users

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/3695)

- [x] This change addresses only certain aspects of the issue

## Description
This creates an endpoint for getting a csv export of users 

As per convo with Sarah we converted this to send the csv as an email

## How Can This Be Tested/Reviewed?
Unit and E2E tests were added, but you can also use http://localhost:3101/api/ to test further

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: backend_new/src/services/user.service.ts [hunk 1]
```diff
@@ -782,6 +783,73 @@ export class UserService {
...
+      return accum;
+    }, {});
+
+    const csvData = buildFromIdIndex(parsedUsers);
+    await this.emailService.sendCSV(
+      requestingUser.jurisdictions,
+      requestingUser,
+      csvData,
+      'User Export',
+      'a user export',
```
Comment (ludtkemorgan) [inline backend_new/src/services/user.service.ts:846]: This is a bit clunky of wording for the user export. I wonder if it should be "The attached file is an export of all users."

**Review by ludtkemorgan — APPROVED** [review]
🦒

File: backend_new/src/services/email.service.ts [hunk 2]
```diff
@@ -403,4 +421,52 @@ export class EmailService {
       throw new HttpException('email failed', 500);
     }
   }
+
+  /**
+   *
+   * @param jurisdictionIds the set of jurisdicitons for the user (sent as IdDTO[]
+   * @param user the user that should received the csv export
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/services/email.service.ts:428]: I think this is actually correct. The first `user` is the name of the param and then the `the user` portion is explaining what the param is for. Lemme know if thats okay or if you're looking for a rewording of the variable description here

File: backend_new/src/utilities/csv-builder.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    Object.keys(record).forEach((key) => {
+      const val = record[key];
+      const groupKeys = extraGroupKeys && extraGroupKeys(key, initialKeys);
+      if (extraHeaders && extraHeaders[key] && groupKeys) {
+        const ids = Object.keys(val);
+        if (groupKeys.nested && ids.length) {
+          Object.keys(val).forEach((sub_id, i) => {
+            const headerGroup = `${key} (${i + 1})`;
+            groupKeys.keys.forEach((groupKey) => {
+              const column = headerIndex[`${headerGroup} ${groupKey}`];
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend_new/src/utilities/csv-builder.ts:79]: Hard agree, I think when we have time or if we have tests that covers the entire use case of the function then we can look at a refactor

as I add more exports the tests should cover those so should be a relatively smooth transition!

File: backend_new/src/services/user.service.ts (hunk 1 above)
Comment (YazeedLoonat) [reply backend_new/src/services/user.service.ts:846]: nice call out, updated!

**Pull request merged**